### ISSUE TITLE
Lost exception info (and exception type) across ABI boundary

### ISSUE THREAD

**Issue by Sergio0694** [microsoft/CsWinRT#1393]
**Describe the bug**

I have a managed WinRT type, [`CanvasEffect`](https://github.com/Sergio0694/ComputeSharp/blob/4f05a70177ce1e1fdcf44c96b8d99dd6c7c59f65/src/ComputeSharp.D2D1.WinUI/CanvasEffect.cs#L12) (a class implementing the `ICanvasImage` WinRT interface from Win2D), that also implements a classic COM interface, [`ICanvasImageInterop`](https://github.com/Sergio0694/ComputeSharp/blob/4f05a70177ce1e1fdcf44c96b8d99dd6c7c59f65/src/ComputeSharp.D2D1.WinUI/ABI.Microsoft.Graphics.Canvas/ICanvasImageInterop.cs#L87). This type can thrown an `InvalidOperationException` ([here](https://github.com/Sergio0694/ComputeSharp/blob/4f05a70177ce1e1fdcf44c96b8d99dd6c7c59f65/src/ComputeSharp.D2D1.WinUI/CanvasEffect.Interop.cs#L91)) from one of the implemented interface methods. CsWinRT correctly handles this in the marshalling stub, catches the exception and calls `ExceptionHelpers.SetExceptionInfo`, but on the other side it seems like some information is lost. Specifically, while I am seeing the same HRESULT being thrown (`-2146233079`), I'm getting a `COMException`, rather than `InvalidOperationException`.

Note: that HRESULT is `COR_E_INVALIDOPERATION`, defined [here](https://github.com/dotnet/runtime/blob/57085686bf7223e2305720383f4e00695321919a/src/libraries/Common/src/System/HResults.cs#L65).

I noticed this because I have a unit test, [`CanvasEffect_EffectNotRegisteringAnOutputNode`](https://github.com/Sergio0694/ComputeSharp/blob/4f05a70177ce1e1fdcf44c96b8d99dd6c7c59f65/tests/ComputeSharp.D2D1.WinUI.Tests/Tests/CanvasEffectTests.cs#L94), which is specifically checking for that exception type, and I noticed that while it used to work correctly, it's now failing, as the resulting exception is just a `COMException`.

Note that the original exception message is not present in the exception data either, it just seems to be lost:

![image](https://github.com/microsoft/CsWinRT/assets/10199417/7ca74be2-0cb4-459b-9cea-d75e6ee2d48b)

**To Reproduce**

1. Clone https://github.com/Sergio0694/ComputeSharp
2. Run the `CanvasEffect_EffectNotRegisteringAnOutputNode` test

**Expected behavior**

The test should (presumably?) re-throw the `InvalidOperationException`, ideally with the original message too.

**Version Info**

- Tested with .NET 7 and .NET 8, same behavior
- CsWinRT 2.0.3 and 2.0.4, same behavior
- Win2D 1.1.0

**Comment by manodasanW** [comment]
We might need to add that error code to the mapping [here](https://github.com/microsoft/CsWinRT/blob/71169932920500310921b4c3fadba1e5a99c6138/src/WinRT.Runtime/ExceptionHelpers.cs#L216C1-L223C1) and also look at why we aren't bringing back the old exception from the error info itself or why it is getting lost.  I assume none of this is cross processes and is within the same process?

**Comment by Sergio0694** [comment]
> "I assume none of this is cross processes and is within the same process?"

Yup! This is all in-proc, it's just a simple app using Win2D and doing interop with some custom types.

I can put together a simpler repo if it helps, unless just reproing from that repo directly is ok, let me know what you prefer 🙂

**Comment by jlaanstra** [comment]
> > "I assume none of this is cross processes and is within the same process?"
> 
> Yup! This is all in-proc, it's just a simple app using Win2D and doing interop with some custom types.
> 
> I can put together a simpler repo if it helps, unless just reproing from that repo directly is ok, let me know what you prefer 🙂

Small repro is always useful I think?

**Comment by jlaanstra** [comment]
Did it work before #1335?

**Comment by dongle-the-gadget** [comment]
I feel like to fix this, we would either have to:
- Copy the `GetExceptionForHR` implementation from .NET, add our own `message`
- Request .NET to add a `message` parameter to `GetExceptionForHR`

This is since C#/WinRT may add its own message to the exception.

**Comment by manodasanW** [comment]
Tried this out with latest build and this seems to be fixed.
